### PR TITLE
fix: built-in function: support char_length() and character_length() function

### LINKED ISSUE #818: built-in function: Support `char_length()` and `character_length()` function
**Is your feature request related to a problem? Please describe.**

### char_length(expr) 
Returns the character length of string data or number of bytes of binary data. The length of string data includes the trailing spaces. The length of binary data includes binary zeros.

### Examples:
```sql
> SELECT char_length('Spark SQL ');
 10
> SELECT CHAR_LENGTH('Spark SQL ');
 10
> SELECT character_length('Spark SQL ');
 10
> SELECT CHARACTER_LENGTH('Spark SQL ');
 10
```

### Reference:
https://spark.apache.org/docs/2.3.0/api/sql/#char_length

**Describe the solution you'd like**

We do provide a built-in function development guide for you. 
Please check [OpenMLDB Build-In Function Develop Guide](https://github.com/4paradigm/OpenMLDB/blob/main/docs/en/built_in_function_develop_guide.md) or the [Chinese version](https://github.com/4paradigm/OpenMLDB/blob/main/docs/cn/built_in_function_develop_guide.md) for help.

Please make sure you:
- [ ] Add built-in C++ function in src/udf/udf.h and src/udf/udf.cc
- [ ] Register function to default library from function `void DefaultUdfLibrary::IniStringUdf()` in src/udf/default_udf_library.cc.
- [ ] Add related unit test in src/codegen/udf_ir_builder_test.cc
- [ ] Documenting function 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Ivyee17**
close #818

### CODE REVIEW

**Comment by Ivyee17** [comment]
The reason of coverage decreasing is a check of null in stringref. I do think that it’s a better choice to add it in the code even if it is not coveraged.

File: hybridse/src/udf/default_udf_library.cc [hunk 1]
```diff
@@ -895,6 +895,19 @@ void DefaultUdfLibrary::InitStringUdf() {
...
+        .doc(R"(
+            @brief Returns the length of the string.
+
+            Example:
+
+            @code{.sql}
+                SELECT CHAR_LENGTH('Spark SQL ');
+                --output 10
+            @endcode
+            @since 0.5.0)");
```
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc]: 0.6.0

File: hybridse/src/udf/default_udf_library.cc [hunk 2]
```diff
@@ -895,6 +895,19 @@ void DefaultUdfLibrary::InitStringUdf() {
             @since 0.4.0)");
     RegisterAlias("lower", "lcase");
     RegisterAlias("upper", "ucase");
+    RegisterExternal("char_length")
+        .args<StringRef>(static_cast<int32_t (*)(StringRef*)>(udf::v1::char_length))
+        .doc(R"(
+            @brief Returns the length of the string.
```
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc]: pls add more note: 
1. measured in characters
2. do not support multibyte character string

File: hybridse/src/udf/default_udf_library.cc [hunk 3]
```diff
@@ -895,6 +895,19 @@ void DefaultUdfLibrary::InitStringUdf() {
...
+
+            Example:
+
+            @code{.sql}
+                SELECT CHAR_LENGTH('Spark SQL ');
+                --output 10
+            @endcode
+            @since 0.5.0)");
+    RegisterAlias("character_length","char_length");
```
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc]: styling

File: hybridse/src/udf/udf.h [hunk 4]
```diff
@@ -271,6 +273,9 @@ void lcase(StringRef *str, StringRef *output, bool *is_null_ptr);
 void ucase(StringRef *str, StringRef *output, bool *is_null_ptr);
 void init_udfcontext(UDFContext* context);
 void trivial_fun();
+
```
Comment (aceforeverd) [inline hybridse/src/udf/udf.h]: extra line ?

File: hybridse/src/udf/udf.cc [hunk 5]
```diff
@@ -157,6 +157,15 @@ int32_t weekofyear(Date *date) {
     }
 }
 
+int32_t char_length(StringRef *str){
+    if (nullptr == str) {
+        return 0;
+    }
+    std::string words = str->ToString();
+    int32_t res = words.length();
+    return res;
```
Comment (aceforeverd) [inline hybridse/src/udf/udf.cc:165]: just `str->size_`

File: hybridse/src/codegen/udf_ir_builder_test.cc [hunk 6]
```diff
@@ -1007,6 +1007,14 @@ TEST_F(UdfIRBuilderTest, degrees) {
     CheckUdf<double, double>(udf_name, -90.0, -pi/2);
     CheckUdf<Nullable<double>, Nullable<double>>(udf_name, nullptr, nullptr);
 }
+
+TEST_F(UdfIRBuilderTest, char_length_udf_test) {
+    auto udf_name = "char_length";
+    CheckUdf<int32_t, StringRef>(udf_name, 10, StringRef("Spark SQL "));
```
Comment (aceforeverd) [inline hybridse/src/codegen/udf_ir_builder_test.cc:1012]: ensure special char like `\n` is correctly counted

**Review by aceforeverd — CHANGES_REQUESTED** [review]

**Comment by aceforeverd** [comment]
> The reason of coverage decreasing is a check of null in stringref. I do think that it’s a better choice to add it in the code even if it is not coveraged.

It's ok

File: hybridse/src/udf/default_udf_library.cc (hunk 1 above)
Comment (Ivyee17) [reply hybridse/src/udf/default_udf_library.cc]: ok

File: hybridse/src/udf/default_udf_library.cc (hunk 2 above)
Comment (Ivyee17) [reply hybridse/src/udf/default_udf_library.cc]: ok

File: hybridse/src/udf/default_udf_library.cc (hunk 3 above)
Comment (Ivyee17) [reply hybridse/src/udf/default_udf_library.cc]: ok

File: hybridse/src/udf/udf.h (hunk 4 above)
Comment (Ivyee17) [reply hybridse/src/udf/udf.h]: ok

File: hybridse/src/udf/udf.cc (hunk 5 above)
Comment (Ivyee17) [reply hybridse/src/udf/udf.cc:165]: ok

File: hybridse/src/codegen/udf_ir_builder_test.cc (hunk 6 above)
Comment (Ivyee17) [reply hybridse/src/codegen/udf_ir_builder_test.cc:1012]: ok

File: hybridse/src/udf/default_udf_library.cc [hunk 7]
```diff
@@ -895,6 +895,19 @@ void DefaultUdfLibrary::InitStringUdf() {
...
+            @brief Returns the length of the string. It is measured in characters and multibyte character string is not supported.
+
+            Example:
+
+            @code{.sql}
+                SELECT CHAR_LENGTH('Spark SQL ');
+                --output 10
+            @endcode
+            @since 0.6.0)");
+    RegisterAlias("character_length", "char_length");
```
Comment (zhanghaohit) [inline hybridse/src/udf/default_udf_library.cc]: pls fix all the cpplint warnings

File: hybridse/src/udf/udf.cc [hunk 8]
```diff
@@ -157,6 +157,14 @@ int32_t weekofyear(Date *date) {
     }
 }
 
+int32_t char_length(StringRef *str){
```
Comment (zhanghaohit) [inline hybridse/src/udf/udf.cc]: cpplint warning

**Review by zhanghaohit — CHANGES_REQUESTED** [review]

**Review by zhanghaohit — APPROVED** [review]
LGTM

**Review by aceforeverd — APPROVED** [review]

**Comment by aceforeverd** [comment]
Pls resolve the merge conflicts

**Comment by Ivyee17** [comment]
Oops.. still the coverage problem..

> > The reason of coverage decreasing is a check of null in stringref. I do think that it’s a better choice to add it in the code even if it is not coveraged.
> 
> It's ok

**Pull request merged**